homeassistant/components/mqtt/client.py

```diff
@@ -503,6 +504,7 @@ async def async_subscribe(
         if not isinstance(topic, str):
             raise HomeAssistantError("Topic needs to be a string!")
 
+        already_active_subscription = self._is_active_subscription(topic)
```

Comment:
Good call out - I can make that only relevant for simple topics.  Though wildcards seem fairly rare.

Reply:
Config topics are received fro wildcard subscriptions, they are not rare.

Reply:
No, I more meant the number of actual wildcard subscriptions is low, compared to simple subscriptions.  Not that the use of wildcards is rare.

In my environment, it subscribes to less than half a dozen wildcard topics, and doesn't re-subscribe to them over and over.  As opposed to the simple ones.

But as I said - this is a simple and appropriate fix.  I'll correct it.

Reply:
Subscriptions should have the configured QoS. A published value received on the subscription should be checked on a retain status. The first time it is received all processed subscriptions should be replayed initially. When new subscription requests arrive and a topic was received for it e.g. a config. It can be that some devices use retain for their config and others don't. So to be able to match replaying we should  not check the subscription topic but the topics that we have received a retained payload on, and only store those payload.
This would mean we also need to index on the topics that had retained content. We can do that when a payload is received and retained is set.
Then we should make sure to clean this all up when a topic is cleared from the broker.

Reply:
As @emontnemery mentioned - according to the spec.  the RETAIN flag for received data has no baring on what the RETAIN flag was set to when it is published.

On publish - RETAIN means "Keep this message, expiring any previously RETAINed message.  And send this message for new subscriptions."
On receive - RETAIN means "This message was sent to you because you just subscribed."
This means, after initial subscription of any topic, all future messages, whether published with RETAIN=0 or RETAIN=1, will have RETAIN=0 set in the message on the receiving side.

That's the crux of the problem Erik has with this MR, and why I mentioned 3 possible ways to think about solving that in the context of this MR. There is no way to know which messages we should store as a 'last message'.

This MR essentially treats all messages received as having been published with RETAIN=1, which is not correct, but does allow us to short circuit the necessity to send 8000+ subscription requests for the same topic.  But I need feedback on the possible 3 approaches I listed in the other thread to try and resolve this in a better way.  Each of which has downsides.  Though I also think that this MR, even as is, is such an improvement over the current status quo, the retain issue could be fixed subsequently too.

Now I could only store a message as the 'last message' if it has RETAIN=1, and then have a TTL, and if the TTL has expired, allow the subscription to go through to find out what the current RETAIN value should be - this ends up being "solution" 3 I listed above.

But QoS doesn't really affect RETAIN from what I can tell.  QOS just says  0 =  "We may not get all non-RETAIN=1 messages", 1 = "We will get every message, possibly more than once" and 2 = "We will get every message, exactly once."   The higher QoS coming with greater complexity/latency (as it requires a 0-way, 1-way or 2-way handshake).